File: src/main/java/ash/java/graphql/data/SearchDaoImpl.java
package ash.java.graphql.data;

import ash.kotlin.graphql.data.SearchDao;
import ash.kotlin.graphql.types.movie.MovieType;
import ash.kotlin.graphql.types.multisearch.PersonType;
import ash.kotlin.graphql.types.multisearch.TvShowType;
import com.google.gson.*;
import com.google.gson.reflect.TypeToken;
import com.mashape.unirest.http.HttpResponse;
import com.mashape.unirest.http.JsonNode;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Service;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class SearchDaoImpl implements SearchDao {

    private static final String RESULTS = "results";
    private Gson gson = new Gson();
    private JsonParser parser = new JsonParser();
    private Type movieSearchListType = new TypeToken<List<MovieType>>(){}.getType();

    @NotNull
    @Override
    public List<MovieType> searchMoviesWithQuery(@NotNull String query) {
        Map<String, Object> queryMap = new HashMap<>();
        queryMap.put("query", query);
        HttpResponse<JsonNode> response = TmdbHttpUtils.sendRequest(TmdbUrls.TmdbQueryUrl.MOVIE_SEARCH_URL, queryMap);

        String searchResults = response.getBody().getObject().get(RESULTS).toString();

        return gson.fromJson(searchResults, movieSearchListType);
    }

    @NotNull
    @Override
    public List<MovieType> searchMoviesWithMultipleParameters(@NotNull Map<String, Object> params) {
        HttpResponse<JsonNode> response = TmdbHttpUtils.sendRequest(TmdbUrls.TmdbQueryUrl.MOVIE_SEARCH_URL, params);

        String searchResults = response.getBody().getObject().get(RESULTS).toString();

        return gson.fromJson(searchResults, movieSearchListType);
    }

    @NotNull
    @Override
    public List<Object> searchMultiSearch(@NotNull Map<String, Object> params) {
        List<Object> results = new ArrayList<>();
        HttpResponse<JsonNode> response = TmdbHttpUtils.sendRequest(TmdbUrls.TmdbQueryUrl.MULTI_SEARCH_URL, params);

        JsonArray multiSearchResults = parser.parse(response.getBody().toString())
                .getAsJsonObject()
                .get(RESULTS)
                .getAsJsonArray();

        for (JsonElement element : multiSearchResults) {
            JsonObject jsonObject = element.getAsJsonObject();
            String mediaType = jsonObject.get("media_type").getAsString();

            if (mediaType.equals("movie")) {
                results.add(gson.fromJson(element, MovieType.class));
            } else if (mediaType.equals("tv")) {
                results.add(gson.fromJson(element, TvShowType.class));
            } else if (mediaType.equals("person")) {
                results.add(gson.fromJson(element, PersonType.class));
            }
        }

        return results;
    }
}

File: src/main/kotlin/ash/kotlin/graphql/data/SearchDaoImpl.kt
package ash.kotlin.graphql.data

import ash.kotlin.graphql.types.movie.MovieType
import ash.kotlin.graphql.types.multisearch.PersonType
import ash.kotlin.graphql.types.multisearch.TvShowType
import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonParser
import com.google.gson.reflect.TypeToken

class SearchDaoImpl : SearchDao
{

    private val RESULTS = "results"
    private val gson = Gson()
    private val parser = JsonParser()
    private val movieSearchListType = object : TypeToken<List<MovieType>>() {}.type

    override fun searchMoviesWithQuery(query: String): List<MovieType>
    {
        val queryMap: MutableMap<String, Any> = HashMap()
        queryMap.put("query", query)
        val response = TmdbRequest().sendQueryParamRequest(TmdbQueryUrl.MOVIE_SEARCH_URL, queryMap)

        val searchResults = response.body.`object`[RESULTS].toString()

        return gson.fromJson(searchResults, movieSearchListType)
    }

    override fun searchMoviesWithMultipleParameters(params: Map<String, Any>): List<MovieType>
    {
        val response = TmdbRequest().sendQueryParamRequest(TmdbQueryUrl.MOVIE_SEARCH_URL, params)
        val searchResults = response.body.`object`[RESULTS].toString()

        return gson.fromJson(searchResults, movieSearchListType)
    }

    override fun searchMultiSearch(params: Map<String, Any>): List<Any>
    {
        val results: MutableList<Any> = ArrayList()
        val response = TmdbRequest().sendQueryParamRequest(TmdbQueryUrl.MULTI_SEARCH_URL, params)

        val multiSearchResults: JsonArray = parser.parse(response.body.toString()).asJsonObject[RESULTS].asJsonArray

        multiSearchResults.forEach {
            val jsonObject = it.asJsonObject
            val mediaType = jsonObject["media_type"].asString

            when (mediaType)
            {
                "movie" -> results.add(gson.fromJson(it, MovieType::class.java))
                "tv" -> results.add(gson.fromJson(it, TvShowType::class.java))
                "person" -> results.add(gson.fromJson(it, PersonType::class.java))
            }
        }

        return results
    }
}